adapters/adform/adform.go

```diff
@@ -0,0 +1,380 @@
+			Price:             bid.Price,
+			Adm:               bid.Banner,
+			Width:             bid.Width,
+			Height:            bid.Height,
+			DealId:            bid.DealId,
+			CreativeMediaType: string(openrtb_ext.BidTypeBanner),
+		}
+
+		bids = append(bids, &pbsBid)
+	}
+
+	return bids
+}
+
+// COMMON
+
+func (r *adformRequest) buildAdformUrl(a *AdformAdapter) string {
+	adUnitsParams := make([]string, 0, len(r.adUnits))
+	for _, adUnit := range r.adUnits {
+		str := fmt.Sprintf("mid=%s", adUnit.MasterTagId)
+		adUnitsParams = append(adUnitsParams, base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString([]byte(str)))
+	}
+	uri := a.URI
+	if r.isSecure {
+		uri = strings.Replace(uri, "http://", "https://", 1)
+	}
+	return fmt.Sprintf("%s/?CC=1&rp=4&fd=1&stid=%s&%s", uri, r.tid, strings.Join(adUnitsParams, "&"))
+}
+
+func (r *adformRequest) buildAdformHeaders(a *AdformAdapter) http.Header {
+	header := http.Header{}
+
+	header.Set("Content-Type", "application/json;charset=utf-8")
+	header.Set("Accept", "application/json")
+	header.Set("X-Request-Agent", fmt.Sprintf("PrebidAdapter %s", a.version))
+	header.Set("User-Agent", r.userAgent)
+	header.Set("X-Forwarded-For", r.ip)
+	if r.referer != "" {
+		header.Set("Referer", r.referer)
+	}
+	if r.userId != "" {
+		header.Set("Cookie", fmt.Sprintf("uid=%s", r.userId))
+	}
+
+	return header
+}
+
+func parseAdformBids(response []byte) ([]*adformBid, error) {
+	var bids []*adformBid
+	if err := json.Unmarshal(response, &bids); err != nil {
+		return nil, err
+	}
+
+	return bids, nil
+}
+
+// BIDDER Interface
+
+func NewAdformBidder(client *http.Client, endpointURL string, usersyncURL string, externalURL string) *AdformAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+
+	redirectUri := fmt.Sprintf("%s/setuid?bidder=adform&uid=$UID", externalURL)
+
+	info := &pbs.UsersyncInfo{
+		URL:         fmt.Sprintf("%s%s", usersyncURL, url.QueryEscape(redirectUri)),
+		Type:        "redirect",
+		SupportCORS: false,
+	}
+
+	return &AdformAdapter{
+		http:         a,
+		URI:          endpointURL,
+		usersyncInfo: info,
+		version:      "0.1.0",
+	}
+}
+
+func (a *AdformAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	adformRequest, errors := openRtbToAdformRequest(request)
+	if len(adformRequest.adUnits) == 0 {
+		return nil, errors
+	}
+
+	requestData := adapters.RequestData{
+		Method:  "GET",
+		Uri:     adformRequest.buildAdformUrl(a),
+		Body:    nil,
+		Headers: adformRequest.buildAdformHeaders(a),
+	}
+
+	requests := []*adapters.RequestData{&requestData}
+
+	return requests, errors
+}
+
+func openRtbToAdformRequest(request *openrtb.BidRequest) (*adformRequest, []error) {
+	adUnits := make([]*adformAdUnit, 0, len(request.Imp))
+	errors := make([]error, 0, len(request.Imp))
+	for _, imp := range request.Imp {
+		// adapter supports only banner impressions for now.
+		if imp.Native != nil || imp.Audio != nil || imp.Video != nil {
```

Comment:
Technically the spec allows for requesting multiple types of ads for one Imp. It would then be up to the publisher to sort out how to display the ad depending on what type of ad was returned. But still a valid strategy to abstain from bidding if competing against the other types, as they are likely to be bidding higher anyway.

More importantly, some future addition to the spec may allow for yet another ad type. You would be more future proof if you checked for `banner` as a requested type, rather than simply assuming if its not one of the others it must be `banner`.

Reply:
Fixed